File: leetcode/combinations.py
# Given two integers n and k, return all possible combinations of k numbers 
# out of 1 ... n.
# Time Complexity: O(n^k)
# Space Complexity: O(n^k)

class Solution:
    # @param {integer} n
    # @param {integer} k
    # @return {integer[][]}
    def combine(self, n, k):
        result, x = [], (1 << k) - 1
        while x < (1 << n):
            result.append([i + 1 for i in xrange(n) if ((x >> i) & 1)])
            u = x & -x
            v = u + x
            x = v | (((v^x) / u) >> 2)
        return result

if __name__ == "__main__":
    solution = Solution()
    print solution.combine(4, 2)

File: leetcode/combinations.js
/* Given two integers n and k, return all possible combinations of k numbers 
 * out of 1 ... n.
 * Time Complexity: O(n^k)
 * Space Complexity: O(n^k)
 */

/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
var combine = function(n, k) {
    var result = [];
    for (var x = (1 << k) - 1; x < (1 << n); ) {
        var temp = [];
        for (var i = 0; i < n; ++i)
            if ((x >> i) & 1)
                temp.push(i + 1);
        result.push(temp);
        u = x & -x;
        v = u + x;
        x = v | (((v^x) / u) >> 2);
    }
    return result;
};
